Update Target Data for updated data standards (#208)
hubverse-org/hubDocs: docs/source/user-guide/target-data.md

each time point (e.g., cases, hospitalizations, and deaths), the values
of those signals will be given in different rows, with a column such as
`signal` indicating what quantity is reported in each row. The only
restrictions that hubverse tools impose on data in this format is that
it should have a column named `observation` and a column with a time
index, such as `date` or `time`.

### Oracle output

Oracle output follows a format that is similar to a [hubverse model
output
file](https://hubverse.io/en/latest/user-guide/model-output.html#example-model-submission-file),
with three main differences:

- Predictions correspond to a distribution that places probability 1 on
  the observed target outcome (see figure below).
- Predictions (e.g., means, quantile values, or pmf category
  probabilities, etc.) are stored in a column named `oracle_value`
  rather than `value`.
- We introduce some conventions to avoid duplication of data, described
  in more detail below. Generally, the columns of the oracle output will
  be a subset of the columns of valid model output for the hub, with
  just those columns that are needed to correctly align `oracle_value`s

>   with the corresponding predicted `value`s produced by modelers.

Author: Good catch! Thank you!

| 0 | 25 | 2022-11-19 | wk inc flu hosp | quantile | 0.9 | 71 |
| 0 | 25 | 2022-11-19 | wk inc flu hosp | quantile | 0.95 | 80 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.05 | 5 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.1 | 21 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.25 | 38 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.5 | 51 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.75 | 64 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.9 | 81 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.95 | 97 |

A hub will typically have access to data in time series format, and will
need to convert it to the **oracle output** format for use with any tools
that require it in that format (see the next section). In hubs that
collect mean, median, quantile, or sample predictions for the reported
signal values in the raw time series data, the two formats may be
essentially the same, perhaps with some renaming of columns. However, in
hubs that form predictions for quantities that are derived from the the
raw time series data, such as the peak time or peak incidence, and in

> hubs that collect pmf or cdf predictions, the formats will differ more

Reviewer: Should we add one or two sentences to indicate more information below if available and link it to you? 
Or reorganize the file to have the "more about" directly after?

Author: Agreed this is a bit hedging, so I've moved it before the description of the column types so that it flows better.

indicates which data formats are most commonly used for each purpose.

## How hubs should provide access to target time series data and oracle output

Hubs should ensure that standardized procedures for accessing target
data are available. The data formats that a hub provides may depend on
the needs of the specific hub, and which hubverse tools the hub wants to
use. For example, a hub that will not be conducting evaluations by
comparing predictions to observed target values may not need to provide
data in the oracle output format.

Access to target time series data and oracle output can be provided in
either of two ways:

1.  by providing example code for accessing target time series data
    and/or oracle output programmatically
2.  by storing snapshots of the target time series data and/or oracle
    output in the hub repository

Following general conventions for storage of code related to modeling
hubs, we recommend that any code for data access be provided in a
separate repository following standard language-specific packaging
guidelines, or if the code is small in scope it can be placed within the
`src` folder of the hub’s repository.

> 

Reviewer: Should we add recommendation on where to store the data? For example in a `target-data` folder?

Author: Yes! Thank you for mentioning this! In fact, we should have recommendations for standardized file names (unless we want to have a way to specify this in the admin schema).

`output_type`:

- **For the `mean`, `median`, `quantile`, and `sample` output types, the
  `oracle_value` is the observed value of the prediction target.** This
  `oracle_value` is the same for all quantile levels and all sample
  indices, since a predictive distribution that places all of its
  probability on the observed outcome will have all quantiles equal to
  that value and all samples from that distribution will be equal to the
  observed value.
- **For `pmf` and `cdf` output types, the `oracle_value` is either `1` or `0`**
    - For the `pmf` output type, the `oracle_value` is `1` when the
      `output_type_id` corresponds to the observed category (indicating a
      probability of 1 for that category) and `0` for other categories.
    - For the `cdf` output type, the `oracle_value` is `0` for any
      `output_type_id` levels that are less than the observed value, and `1`
      for any `output_type_id` levels that are greater than or equal to the
      observed value, corresponding to the step function cdf of a
      probability distribution that places all of its probability at the
      observed value.

> ### Oracle output vs model output

Reviewer: Should we put the information a little bit higher, I think it helps a lot with understanding why we design it that way. For example, at the beginning of the "Oracle output" section.

Author: I think that's a good point. I'll move it up.

difference between the outputs is that the oracle output is necessarily going
to have a subset of the task ID columns as the model output data and, depending
on the hub, may not have either of the model output representation columns.

#### Model output representation columns

**The `output_type` and `output_type_id` columns only need to be included if
the hub collects `pmf` or `cdf` outputs.** As was described above, for those
two output types the `oracle_value` depends on the `output_type_id`. On the
other hand, the `oracle_value` is not specific to the quantile level for
quantile forecasts or the sample index for sample forecasts, and so for these
output types (as well as mean and median), the `output_type_id` is not needed
to align observations with predictions.

#### Task ID columns

The model output can contain any number of task ID columns that are used to
provide details about what is being predicted based on

 - independent task ID variables (e.g. `location`, `target_date`, and
   `age_group`),
 - derived (dependent on other variables) task ID variables (e.g. `horizon`,
   `reference_date`, `origin_date`), and
 - scenario-specific task ID variables (e.g. `scenario_id`).

> **The oracle output will contain the independent task ID variables** that are

Reviewer: Do you mean "independent and any other variables" necessary to match?
Because you will need some date/time information probably and with this list, it seems that all time related variable are called "derived".

Author: Probably. I was trying to avoid over-complicating the statement. Thinking about it, the oracle data should have the same columns as the time series data (except `date` is renamed to `target_date`) with additional columns that describes the metadata about the outputs.... unless I'm completely off-base? 

I was attempting to summarize @elray1's original text:

>  - The oracle output should include enough of the task id variables and columns with metadata about the outputs (output_type and output_type_id) to uniquely identify which oracle_values correspond to which predicted values. For example, this will typically include task id variables such as location and target_date (or target_end_date), since the oracle_value will be specific to the location and target_date.
>  - Any task id variables that are not necessary to match observations with predictions can be omitted from the oracle output. For example, if target_date is included then reference_date and horizon variables can be omitted because the same observation will generally correspond to a particular target_date regardless of the forecast horizon. Similarly, in a scenario projection setting, the scenario_id can be omitted.

Reviewer: For me the original text is clear enough. I will be ok with keeping it.

Author: The original text feels a lot like legal jargon and I had a hard time understanding what it actually means.

reference date (the columns `model_id` and `reference_date` are omitted
for compactness):

| horizon | location | target_end_date | target | output_type | output_type_id | value |
|---:|:---|:---|:---|:---|:---|---:|
| 0 | 25 | 2022-11-19 | wk inc flu hosp | quantile | 0.05 | 22 |
| 0 | 25 | 2022-11-19 | wk inc flu hosp | quantile | 0.1 | 31 |
| 0 | 25 | 2022-11-19 | wk inc flu hosp | quantile | 0.25 | 45 |
| 0 | 25 | 2022-11-19 | wk inc flu hosp | quantile | 0.5 | 51 |
| 0 | 25 | 2022-11-19 | wk inc flu hosp | quantile | 0.75 | 57 |
| 0 | 25 | 2022-11-19 | wk inc flu hosp | quantile | 0.9 | 71 |
| 0 | 25 | 2022-11-19 | wk inc flu hosp | quantile | 0.95 | 80 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.05 | 5 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.1 | 21 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.25 | 38 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.5 | 51 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.75 | 64 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.9 | 81 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.95 | 97 |

### Generating oracle output data

> A hub will typically have access to data in time series format, and will
> need to convert it to the **oracle output** format for use with any tools
> that require it in that format (see the next section). In hubs that
> collect mean, median, quantile, or sample predictions for the reported
> signal values in the raw time series data, the two formats may be
> essentially the same, perhaps with some renaming of columns. However, in
> hubs that form predictions for quantities that are derived from the the
> raw time series data, such as the peak time or peak incidence, and in
> hubs that collect pmf or cdf predictions, the formats require the presence of
> model output representation columns.

Reviewer 2: ```suggestion
A hub will typically have access to data in time series format, and will
need to convert it to the **oracle output** format for use with any tools
that require it in that format (see the next section). In hubs that
collect mean, median, quantile, or sample predictions for the reported
signal values in the raw time series data, the two formats may be
essentially the same, perhaps with some renaming of columns. However,
these data formats will differ more in hubs that form predictions for quantities
that are derived from the the raw time series data, such as the peak time or
peak incidence, and in hubs that collect pmf or cdf predictions.
```

Reviewer 2: See a separate comment on the PR for more about my thinking here.  I decided that maybe simpler was better in terms of just stating that the formats would differ in these situations rather than providing more detailed examples.  I'm open to further revision though!

Author: I think this is definitely moving in the right direction!

### Task ID columns

**The oracle output should include enough of the task ID variables to uniquely
identify which `oracle_values` correspond to which predicted values.** In the
[above oracle output example](#oracle-intro-example), the `location`,
`target_end_date`, and `target` columns are included because they are necessary
to identify _where_ and _when_ a given _target_ was measured as the
`oracle_value`.

Similarly, **any task ID variables that are not necessary to match observations
with predictions can be omitted from the oracle output.** In the [above oracle
output example](#oracle-intro-example), the `horizon`, `model_id`, and
`reference_date` columns are not included. Both `horizon` and `reference_date`
are related to the `target_end_date` and thus would be redundant. Importantly,
_these task ID variables are not applicable for observed data_---they are used
for describing model-specific parameters about unknown events. Likewise, in a
scenario projection setting, the `scenario_id` can be omitted as there is only
one scenario for an observed event[^quanta].

[^quanta]: just don't tell the quantum physicists.

### Model output representation columns

> **The `output_type` and `output_type_id` columns only need to be included if
> the hub collects `pmf` or `cdf` outputs.** As was described above, for those
> two output types the `oracle_value` depends on the `output_type_id`. On the

Reviewer 2: ```suggestion
**The `output_type` and `output_type_id` columns only need to be included if
the hub collects `pmf` or `cdf` outputs.** For those two output types the
`oracle_value` depends on the `output_type_id` (see the next section for more
detail). On the
```

Author: I think the "for" was a typo in there and can be removed unless we want to have it sound more prophetic than it should.

| wk flu hosp rate | cdf | week rate of hospital admissions with flu per 100,000 population |
| wk flu hosp rate category | pmf | categorical severity level of the hospital admissions rate, with levels ‘low’, ‘moderate’, ‘high’, and ‘very high’ |

Below, we show snippets of the contents of a `model_out_tbl` with
example forecast submissions and the corresponding **oracle output** for
each `output_type`. We highlight two points about these objects:

- The `reference_date` and `horizon` columns are included in the model
  outputs, but they are not included in the oracle output.
- In this example, the **oracle output** for the `mean`, `median`,
  `quantile`, and `sample` output types are all the same, and they
  contain `<NA>` values for the `output_type_id`. In a hub without `pmf`
  or `cdf` output types, the `output_type` and `output_type_id` columns
  could be omitted and this duplication could be eliminated.

:::{note}

These examples are all collected and filtered from [the `hubExamples` package](https://hubverse.github.io/hubExamples). The model output data set contains over
10,000 rows and the oracle output data has over 200,000 rows.

> To make comparisons easier, we have subset the data to Massachusettes (FIPS code
> 25) with four target end dates between 2022-11-19 and 2022-12-10.

Reviewer 2: ```suggestion
To make comparisons easier, we have subset the data to Massachusettes (FIPS code
25) with one `reference_date` of 2022-11-19 and four target end dates between 2022-11-19 and 2022-12-10.
```

| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.9 | 81 |
| 1 | 25 | 2022-11-26 | wk inc flu hosp | quantile | 0.95 | 97 |
:::

:::{table} The `quantile` **oracle output** from 19 November to 10 December 2022
| location | target_end_date | target | output_type | output_type_id | oracle_value |
|:---|:---|:---|:---|:---|---:|
| 25 | 2022-11-19 | wk inc flu hosp | quantile | `<NA>` | 79 |
| 25 | 2022-11-26 | wk inc flu hosp | quantile | `<NA>` | 221 |
| 25 | 2022-12-03 | wk inc flu hosp | quantile | `<NA>` | 446 |
| 25 | 2022-12-10 | wk inc flu hosp | quantile | `<NA>` | 578 |
:::

As with the `mean` and `median` output types, the `oracle_value` for a
quantile type is the observed numeric value of the prediction target,
which is the quantile of a predictive distribution that assigns
probablity 1 to that observed value at any quantile probability level. A
model output file would need to have a separate row for each quantile
level reported in the `output_type_id` column. As a space-saving
convention, we use `output_type_id = <NA>` to indicate that this
`oracle_value` applies to all quantile levels.

### Output type `sample`

> :::{table} A subset of **model output** showing 6 `sample` predictions across one horizon

Reviewer 2: ```suggestion
:::{table} A subset of **model output** showing 6 `sample` predictions at one horizon
```

| 2 | 25 | 2022-12-03 | wk flu hosp rate category | pmf | high | 0.0000000 |
| 2 | 25 | 2022-12-03 | wk flu hosp rate category | pmf | very high | 0.0000000 |
:::

:::{table} The `pmf` **oracle output** from 19 November to 03 December 2022
| location | target_end_date | target | output_type | output_type_id | oracle_value |
|:---|:---|:---|:---|:---|---:|
| 25 | 2022-11-19 | wk flu hosp rate category | pmf | low | 1 |
| 25 | 2022-11-19 | wk flu hosp rate category | pmf | moderate | 0 |
| 25 | 2022-11-19 | wk flu hosp rate category | pmf | high | 0 |
| 25 | 2022-11-19 | wk flu hosp rate category | pmf | very high | 0 |
| 25 | 2022-11-26 | wk flu hosp rate category | pmf | low | 0 |
| 25 | 2022-11-26 | wk flu hosp rate category | pmf | moderate | 1 |
| 25 | 2022-11-26 | wk flu hosp rate category | pmf | high | 0 |
| 25 | 2022-11-26 | wk flu hosp rate category | pmf | very high | 0 |
| 25 | 2022-12-03 | wk flu hosp rate category | pmf | low | 0 |
| 25 | 2022-12-03 | wk flu hosp rate category | pmf | moderate | 0 |
| 25 | 2022-12-03 | wk flu hosp rate category | pmf | high | 1 |
| 25 | 2022-12-03 | wk flu hosp rate category | pmf | very high | 0 |
:::

> The presence of a `1` for the `oracle_value` in the first row and `0` in the
> subsequent three rows indicates that the observed rate category in
> Massachusettes on the week of 2022-11-19 was `"low"` while the week of
> 2022-11-26 was `"moderate"`.

Reviewer 2: ```suggestion
The presence of a `1` for the `oracle_value` in the first row and `0` in the
subsequent three rows indicates that the observed rate category in
Massachusettes on the week of 2022-11-19 was `"low"`. Similarly, the observed rate category for the week of
2022-11-26 was `"moderate"`.
```
